### PR TITLE
DFU

### PR DESCRIPTION
**Pull request by NickKibish**
# DFU
New version of iOS-DFU Library was integrated into project. 
DFU workflow was updated.
## How that works:
DFU view controller consists of two parts:
* **DFU Preview**: instance of `DFUFileView` class which shows, schematic preview of Firmware and updating progress
* **Log Text view**: subclass of `UITextView` which conforms `LoggerDelegate` protocol and displays logs from DFULibrary as attributed text.

`DFUFileView` can have few states:
* **ready to open file**: user can chose distribution packet from documents or drag it into view (on iPad)
* **ready for update**: Displays file's size and schematic view of distribution packet. User can select another file or start update progress
* **updating**: shows progress of updating.

IMG_0139
IMG_0138
IMG_0137

DFUFileView also conforms `DFUProgressDelegate` and handls updating progress itself.

### **Nordic Button**: 
branded button with rounded corners. Can have different state and redraw itself on setting new state.
### **ProgressImae**: 
In fact that is view (Not UIImageView subclass). This class loads image and draw it on layer. 

It also can draw the image with multiple colors to represent different parts. 

`ProgressPart` is a structure that represents ratio between colored parts of image. In my case it's used to show scheme of distribution scheme. 

E.g. to show that in distribution package soft device takes 150kb and application takes 30kb we can create two parts and set them to `parts` property.
```swift
let parts = [
    ProgressPart(parts: 150, color: .green),
    ProgressPart(parts: 30, color: .red)
]

fileView.parts = parts
```

The `progress` is `Int` property expected beeing in range 0...100 (values out of range will be cut and ignorred). It represents "filled" or completed part. 

E.g. If progress is 50% the bottom half of image will be filled with colors (descibed)

If we set only one part and 100% progress it will be the same as `UIImageView` with template rendering mode.

### DFUFirmware
`DFUFirmware` was extended and now conforms `NSItemProviderReading` protocol. It was done for supporting drag&drop on iPad.

### `PeripheralScanner` and `ConnectionViewController`
I've updated scanning system and now application scans devices in background and once a second updates the list. 
That is WIP part. And I want to refactor all `BluetoothManager` block.

## 🎥 You saw it, but...
ezgif com-video-to-gif-2

### CODE REVIEW

File: Podfile [hunk 1]
```diff
@@ -1,6 +1,6 @@
 use_frameworks!
 target "nRF Toolbox" do
-    pod 'iOSDFULibrary', '~> 4.5'
+    pod 'iOSDFULibrary'#, '~> 4.5'
```
Comment (philips77) [inline Podfile]: What does it do?

File: nRF Toolbox/Profiles/DeviceFirmwareUpdate/ViewControllers/DFUViewController.swift [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        navigationItem.rightBarButtonItem = connectBtn
+        
+        navigationItem.title = "DFU"
+        fileView.delegate = self
+        fileView.fileDelegate = self
+        fileView.state = .readyToOpen
+    }
+
+    override var peripheralDescription: PeripheralDescription {
+        PeripheralDescription(uuid: nil, services: [.battery])
```
Comment (philips77) [inline nRF Toolbox/Profiles/DeviceFirmwareUpdate/ViewControllers/DFUViewController.swift]: Why battery? What is this?

File: nRF Toolbox/Profiles/DeviceFirmwareUpdate/ViewControllers/DFUViewController.swift [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        fileView.state = .paused
+    }
+    
+    func resume(_ fileView: DFUFileView) {
+        dfuController?.resume()
+        fileView.state = .updating(.softdeviceBootloaderApplication)
+    }
+    
+    func stop(_ fileView: DFUFileView) {
+
```
Comment (philips77) [inline nRF Toolbox/Profiles/DeviceFirmwareUpdate/ViewControllers/DFUViewController.swift]: Empty line

File: nRF Toolbox/Profiles/DeviceFirmwareUpdate/Views/FileScheme/FirmwareProgressImage.swift [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    static let firmwareApplication: UIColor = .nordicRed
+    static let firmwareBootloader: UIColor = .nordicFall
+    static let firmwareSoftDevice: UIColor = .nordicGreen
+}
+
+class FirmwareProgressImage: ProgressImage {
+    func setParts(with firmware: DFUFirmware, reversed: Bool = false ) {
+        var parts = [ProgressPart]()
+        
+        let application = Int(firmware.size.application)
```
Comment (philips77) [inline nRF Toolbox/Profiles/DeviceFirmwareUpdate/Views/FileScheme/FirmwareProgressImage.swift]: In the current DFU lib version for packets with SD and BL (and optionally App) the BL size is set to 1 and the SD size is set to (SD+BL).size - 1.
I think you may have 2 parts: System (SD+BL) and App.

File: nRF Toolbox/Profiles/DeviceFirmwareUpdate/Views/FileSizeView.swift [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    }
+    
+    func update(with firmware: DFUFirmware) {
+        let subviewes = stackView.arrangedSubviews
+        subviewes.forEach {
+            stackView.removeArrangedSubview($0)
+            $0.removeFromSuperview()
+        }
+        
+        let bootloader = Int(firmware.size.bootloader)
```
Comment (philips77) [inline nRF Toolbox/Profiles/DeviceFirmwareUpdate/Views/FileSizeView.swift]: Here same as before.

File: nRF Toolbox/Profiles/DeviceFirmwareUpdate/Views/LogerTextView.swift [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+//  nRF Toolbox
+//
+//  Created by Nick Kibysh on 05/12/2019.
+//  Copyright © 2019 Nordic Semiconductor. All rights reserved.
+//
+
+import UIKit
+import iOSDFULibrary
+
+class LogerTextView: UITextView, LoggerDelegate {
```
Comment (philips77) [inline nRF Toolbox/Profiles/DeviceFirmwareUpdate/Views/LogerTextView.swift]: Logger (double g)

File: nRF Toolbox/Profiles/DeviceFirmwareUpdate/Views/NordicButton.swift [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+class NordicButton: UIButton {
+    
+    struct Style: RawRepresentable, Equatable {
+        let rawValue: Int
+        
+        static let `default` = Style(rawValue: 1)
+        static let mainAction = Style(rawValue: 2)
+        static let distructive = Style(rawValue: 3)
```
Comment (philips77) [inline nRF Toolbox/Profiles/DeviceFirmwareUpdate/Views/NordicButton.swift]: destructive

File: nRF Toolbox/Profiles/Template/PeripheralViewController.swift [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        Log(category: .ble, type: .debug).log(message: "Changed Bluetooth status in \(String(describing: type(of: self))), status: \(status)")
+        switch status {
+        case .poweredOff:
+            activePeripheral = nil
+
+            let bSettings: InfoActionView.ButtonSettings = ("Settings", {
+                let url = URL(string: "App-Prefs:root=Bluetooth") //for bluetooth setting
+                let app = UIApplication.shared
+                app.open(url!, options: [:], completionHandler: nil)
+
```
Comment (philips77) [inline nRF Toolbox/Profiles/Template/PeripheralViewController.swift]: Empty line

**Comment by NickKibish** [comment]
###New Logger View
Changed logger view to table view. Now it also shows time.
IMG_F0F7821AE514-1

File: Podfile (hunk 1 above)
Comment (NickKibish) [reply Podfile]: set correct version

File: nRF Toolbox/Profiles/DeviceFirmwareUpdate/Views/FileScheme/FirmwareProgressImage.swift (hunk 4 above)
Comment (NickKibish) [reply nRF Toolbox/Profiles/DeviceFirmwareUpdate/Views/FileScheme/FirmwareProgressImage.swift]: done

File: nRF Toolbox/Profiles/DeviceFirmwareUpdate/Views/FileSizeView.swift (hunk 5 above)
Comment (NickKibish) [reply nRF Toolbox/Profiles/DeviceFirmwareUpdate/Views/FileSizeView.swift]: done

**Pull request merged**